rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -589,6 +601,15 @@ trait SpatialMatcherInstances {
       } else
         MonoidK[F].empty
 
+  implicit def gdeployerAuthSpatialMatcherInstance[F[_]: Splittable: Alternative: Monad: _error: _cost: _freeMap]
+      : SpatialMatcher[F, GDeployerAuth, GDeployerAuth] =
+    (target, pattern) =>
+      if (target == pattern) {
+        val cost = equalityCheckCost(target, pattern)
+        charge[F](cost) *> ().pure[F]
```

Comment:
Charging should be done regardless of the comparison result.

Do other matching rules repeat this error?

Reply:
Yes, other matching rules repeat this.

The reasoning might be that checking for equality on two things that are equal is much more expensive than checking for equality on two things that are different.

Reply:
Maybe there should be a constant small cost for the inequality case.

Reply:
That would be a wrong reasoning, as I can maliciously make the compared things differ at the last bits :)

I've reviewed the usages of equalityCheckCost and this pattern might make some sense in cases where we delegate further to `spatialMatch` when two lists of things are equal (see e.g. ReceiveBind or Match). Spatial match is going to do the charging for each of the elements in that case.

It doesn't make sense for ground terms though (and this includes all GUnforgrable-s)

Reply:
> That would be a wrong reasoning, as I can maliciously make the compared things differ at the last bits :)
I was under the assumption that this is not possible. How would you do that for `GPrivate` or `GDeployerAuth`?

Reply:
You're right, that would require rolling two very similar public keys. That's not the real point though:

> Maybe there should be a constant small cost for the inequality case.

It can't be constant. In case of GDeployerAuth and GPrivate both outcomes of the comparison have the same cost lower bound. Granted, there are cases where the comparison short-circuits after 2 or 3 bytes - or even the first one! - but almost none of them do that without incurring any computation. So we can't be saying "if they're not equal, charge nothing".

So we either charge both outcomes equally, or need to integrate cost accounting into the equals method (at which point it shadows the actual computation cost of the equality).

Reply:
> So we can't be saying "if they're not equal, charge nothing".

But I still think we could say "if they're not equal, charge a small fixed amount" since on average that small fixed amount would reflect the actual cost (and I don't see how it could be an attack vector).
However I'm perfectly fine with the better-safe-than-sorry option for now :)

Reply:
> since on average that small fixed amount would reflect the actual cost

One of the main points of cost accounting is securing against the ones that play **against** the average. We must employ a pessimistic, upper-bound based mindset in this area. It can't be an attack vector only because of the particular case we have here.

Also, we really should rather make the cost model simple and uniform than try to make it "optimal" in terms of how little we charge. That's not the point, and actually against the interests of validators, and - to a certain degree - against the interests of the community / coop / platform itself.